pkg/credentials/iam_aws.go

```diff
@@ -116,11 +116,19 @@ func (m *IAM) Retrieve() (Value, error) {
 			roleSessionName: os.Getenv("AWS_ROLE_SESSION_NAME"),
 		}
 
-		stsWebIdentityCreds, err := creds.Retrieve()
-		if err == nil {
-			m.SetExpiration(creds.Expiration(), DefaultExpiryWindow)
+		// FIXME: a more comprehensive fix is needed for all
+		// providers with an expontential retry and logging.
+		//
+		// Loop until getting a valid credentials
+		for {
+			stsWebIdentityCreds, err := creds.Retrieve()
+			if err == nil {
+				m.SetExpiration(creds.Expiration(), DefaultExpiryWindow)
+				return stsWebIdentityCreds, nil
+			}
+
+			time.Sleep(10 * time.Second)
```

Comment:
We should log the error @vadmeste perhaps by taking some custom logger function here as part of the API such that we can log the errors here.

Reply:
> We should log the error @vadmeste perhaps by taking some custom logger function here as part of the API such that we can log the errors here.

For logging, the only thing that I can think of is something like this:
`SetCustomLogger(fn func(error))`

But also I wanted to send a more comprehensive PR today, https://github.com/minio/minio-go/pull/1563 but I want to discuss it first. Maybe let's merge this first and then I can send another PR with more careful testing.

Reply:
Sure atleast we need logging here @vadmeste